### ISSUE TITLE
Cannot build project depending on serde using Cargo from 2017

### ISSUE THREAD

**Issue by aidanhs** [rust-lang/cargo#14237]
### Problem

I was expecting to be able to build an old project with a cargo.lock and correct rustc, but it fails because of a registry parsing error for an unrelated version of serde

### Steps

```
$ git clone https://github.com/aidanhs/ayzim.git
$ rustup default rustup default nightly-2017-03-03
~/ayzim $ cargo --version
cargo 0.18.0-nightly (19ea423 2017-03-02)
~/ayzim $ cargo build --verbose
error: failed to parse registry's information for: serde

Caused by:
  Feature `derive` depends on `serde_derive` which is not an optional dependency.
Consider adding `optional = true` to the dependency
```

### Possible Solution(s)

I think when trying to parse the registry, my version of cargo should lazily apply validation checks to versions being considered instead of eagerly validating all versions. Unfortunately even doing this doesn't help my really old version of cargo and I don't really want to update the project...I just want to build it!

### Version

```text
$ cargo --version --verbose
cargo 0.18.0-nightly (19ea423 2017-03-02)
release: 0.18.0
commit-hash: 19ea4238114ea94c3fde0b623618e75accd70a66
commit-date: 2017-03-02
```

**Comment by Eh2406** [comment]
You are absolutely correct that that would be a better behavior for cargo to have, in fact that behavior was added way back in version `0.19`. Which is one version off from being able to help you. We work incredibly hard to add new functionality in such a way that old cargos will continue to work, I am very sorry that we failed in this case. The new functionality has been stable for a long time so there's no easy way to back it out. And without a time machine we have no way of changing outdated versions of cargo people are still using.

**Comment by aidanhs** [comment]
Thanks the reply, fair enough. Is it easy to find the oldest nightly containing the cargo with this fix?

**Comment by weihanglo** [comment]
Probably [`cargo-bisect-rustc`](https://github.com/rust-lang/cargo-bisect-rustc) can do the job?

**Comment by YaLTeR** [comment]
I hit this in [circular-queue](https://github.com/YaLTeR/circular-queue) where our MSRV of 1.17 no longer builds due to this (probably hasn't for a while..). To make matters worse, serde isn't even a required dependency at all, but the build fails regardless.
```
$ cargo +1.17 check --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
error: failed to parse registry's information for: serde

Caused by:
  Feature `derive` depends on `serde_derive` which is not an optional dependency.
Consider adding `optional = true` to the dependency
```

I guess the only workaround is to completely remove those dependencies and features from Cargo.toml right before running any old cargo commands?

**Comment by YaLTeR** [comment]
Seems that it's also broken on 1.18 (cargo 0.19.0), but then it starts working on 1.19 (cargo 0.20.0).

**Comment by epage** [comment]
As there isn't anything we can do about Rust 1.18 but we've at least made it so we don't hit this problem in the future, I'm not seeing much we can do about this at this point and so I'm going to close this.  If there is a reason for us to re-evaluate this, let us know!

**Comment by bugaevc** [comment]
Can't you release Cargo 1.18.1 with the relevant fix backported?
